### ISSUE TITLE
Bernstein's Paper

### ISSUE THREAD

**Issue by loretopaulinojr** [loretopaulinojr/CCOMP#8]

**Comment by loretopaulinojr** [comment]
**Analysis Method:**

This passage describes a method developed by the authors for quantifying the robustness of a metabolic network in producing a specific metabolite from different metabolic precursors. The method involves a probabilistic framework inspired by concepts from statistical physics, specifically percolation theory. Percolation theory is a mathematical framework used to study connectivity properties in various networks by randomly adding or removing components (nodes or edges) and assessing network connectivity.

In this context, the authors use a similar approach to assess the robustness of a metabolic network. They assign probabilities (Pin) to metabolites being present in the network, representing environmental conditions or assumptions about metabolite availability. The network's structure is then used to calculate the probability of producing a target metabolite (Pout) based on these input probabilities. The authors sample different sets of input metabolites probabilistically and use a modified version of Flux Balance Analysis (FBA) to evaluate the network's ability to produce the target metabolite.

The authors introduce a novel metric called the producibility metric (PM), which quantifies the biosynthetic capabilities of the metabolic network. PM is calculated based on the relationship between input probabilities (Pin) and the probability of producing the target metabolite (Pout). A producibility curve is generated, and PM is defined as 1 minus the Pin value at which Pout equals 0.5. Larger PM values indicate increased robustness of the network in producing the target metabolite.

The method is implemented using modular functions written in MATLAB and interfaces with the COBRA toolbox, a software package for constraint-based modeling of biological networks. The authors provide the code for their method, making it freely available for others to use and explore. The code repository can be accessed at the provided GitHub link: `https://github.com/yourusername/yourrepository`

**Comment by loretopaulinojr** [comment]
**Why is it 1 - Pin:**

The equation PM = 1 - Pin represents the calculation of the producibility metric (PM) based on the relationship between input probabilities (Pin) and the probability of producing the target metabolite (Pout). Let's break it down to understand why it is 1 minus Pin:

1. **Producibility Curve:** The authors generate a producibility curve, which describes how the probability of producing the target metabolite (Pout) varies with different input probabilities (Pin). This curve represents the relationship between the availability of input metabolites (Pin) and the likelihood of successfully producing the desired metabolite (Pout).

2. **Pin Value at Pout = 0.5:** The authors are interested in finding the point on the producibility curve where the probability of producing the target metabolite is 0.5. This means there is an equal chance of producing or not producing the metabolite given the specified input probabilities.

3. **Definition of PM:** PM is defined as 1 minus the Pin value at which Pout is equal to 0.5. In other words, PM represents the input probability (Pin) at the threshold where the network has a 50% chance of producing the target metabolite.

4. **Interpretation:** A larger PM value indicates increased robustness of the metabolic network in producing the target metabolite. If PM is close to 1, it means that even with a relatively low probability of having the necessary precursors (Pin), the network can still produce the metabolite with a 50% chance. On the other hand, if PM is closer to 0, it suggests that a higher probability of having the precursors (Pin) is required to achieve a 50% chance of producing the metabolite.

**Comment by loretopaulinojr** [comment]
**In-Depth Interpretation:**

The producibility metric (PM) provides valuable insights into the metabolic network's robustness in producing the target metabolite. Here's an in-depth interpretation of PM:

1. **Larger PM Value Indicates Increased Robustness:**
   - A PM value closer to 1 signifies a highly robust metabolic network.
   - Even with a relatively low probability (Pin) of having the necessary precursors, the network can produce the target metabolite with a 50% chance.
   - High PM values demonstrate the network's adaptability and efficient resource utilization.

2. **PM Close to 0 Requires Higher Probability of Precursors:**
   - A PM value closer to 0 suggests a less robust metabolic network.
   - A higher probability (Pin) of having the necessary precursors is necessary for a 50% chance of producing the metabolite.
   - Low PM values indicate the network's sensitivity to changes in precursor availability.

In summary, PM serves as a measure of the network's ability to cope with varying environmental conditions. A higher PM value signifies a metabolic network that can achieve its goal (producing the target metabolite) even with limited precursor availability, showcasing its robustness. Conversely, a lower PM value indicates that the network is more dependent on the availability of specific precursors, making it less resilient to changes in the environment.

*[This interpretation is part of the analysis provided in the GitHub repository.]*

**Comment by loretopaulinojr** [comment]
# Order of Code

1. **prep_mod:**
   - This function is used to prepare the metabolic network model for analysis. It standardizes the network models, identifies extracellular and intracellular metabolites, adds exchange reactions, and outputs a new network model. It should be used first to ensure that your metabolic network is properly formatted for subsequent analyses.

2. **calc_PM_fit_nonlin:**
   - This function calculates the producibility metric (PM) for a specified metabolic network model and metabolite. It employs an efficient nonlinear fitting technique. After preparing the network with `prep_mod`, you can use this function to calculate PM for a single metabolite.

3. **rand_add:**
   - This function generates a random sample of input metabolites based on the Bernoulli parameter for each input metabolite. It should be used when you need to introduce random variations to input metabolites.

4. **feas:**
   - The `feas` function determines if the production of a given target metabolite set is feasible given the metabolic network model with specified constraints. It uses Flux Balance Analysis (FBA) to assess feasibility. It should be used when you want to check the feasibility of a particular target metabolite set.

5. **prob:**
   - This function utilizes `rand_add` and `feas` to determine the probability of producing the target metabolite. It generates a chosen number of random samples of input metabolites and calculates the probability of producing the target metabolite based on feasibility. It should be used when you want to estimate the probability of producing a target metabolite.

6. **find_PM_mods_mets:**
   - This function is designed to facilitate the parallelization of the PM calculation. It takes a directory of metabolic network models and calculates PM for multiple metabolites in parallel. It relies on the functionalities of `calc_PM_fit_nonlin` and other specifications. Use this function when you need to perform PM calculations for multiple metabolites across multiple network models.

Ensure that the network is properly prepared with `prep_mod` before performing other analyses. The specific order may vary based on your analysis goals and dependencies between functions. Adjust the sequence based on your specific requirements.

**Comment by loretopaulinojr** [comment]
Based on the paper, he used the calc_PM_fit_nonlin for Ecoli from the ([King et al., 2016](https://github.com/users/loretopaulinojr/projects/1/views/1?pane=issue&itemId=44068028)).

**Comment by loretopaulinojr** [comment]
https://elifesciences.org/articles/39733#content

**Comment by loretopaulinojr** [comment]
% David Bernstein

% example_calculate_PM

% This example code goes through the calculation of the producibility
% metric (PM) from Bernstein et al.

% The code contains a step by step procedure for calculating an example PM
% for 2 metabolites on 2 different metabolic network models.

% Commented code is also provided for calculating the PM for 88 different
% essentail biomass components across 456 metabolic network models from 
% the oral microbiome. However, we do not reccomend running this code
% without parallelizing.

% 1) 
% Initialize the COBRA toolbox and set solver
% The COBRA toolbox is a popular compendium of metabolic network analysis
% methods. It contains the functions for running flux balance analysis
% (FBA) which are used to calculate the PM. Please visit the COBRA toolbox
% webpage for additional information on the toolbox including installation
% (https://opencobra.github.io/cobratoolbox/stable/)
initCobraToolbox
changeCobraSolver('glpk')

% 2) 
% Add to the MATLAB path the folder containing the algorithm funcitons
addpath('./algorithm_functions')

% 3)
% Set Parameters
% Define the parameters for the PM calculation: all parameters are
% described in the comments for the find_PM_mods_mets command which is also
% coppied here:
% INPUT
% results_location: char, directory where final results will be saved
% models_location: char, directory containing model files
% mets: cell array - char, metabolite names that will be analyzed as target
% metabolites 
% metnum: int, index of target metabolite to analyze, used for
% parallelization 
% modnum: int, index of model to analyze, used for parallelization
% fixon: cell array - char, metabolite names that will always be added
% fixoff: cell array - char, metabolite names that will never be added
% a: char, indicator variable for metabolites randomly added during
% producibility analysis, a = int_nt: add intracellular metabolites (remove
% target), a = ext_nt: add extracellular metabolites (remove target), a = int: add
% intracellular metabolites (don't remove target), a = all_nt: add all
% metabolites (remove target)
% s: logical, Boolean variable for mass balance constraint on all
% metabolites, s = 0: equality mass balance constraint (S*V=0), s = 1:
% inequality mass balance constraints (S*V>=0)  
% samp: int, number of samples taken to estimate probability of production
% thresh: double, threshold variable used by nonlinear fitting
% n: int, number variable used by nonlinear fitting
% noise: double, noise variable used by nonlinear fitting
% runs: int, number of repeated runs used to estimate producibility metric
results_location = './example_results';
models_location = './example_models';
mets = {'cpd00118_c0','cpd00156_c0'}; % Putrescine, L-Valine
metnum = 1:2;
modnum = 1:2;
fixon = {};
fixoff = {};
a = 'int_nt';
s = 1;
samp = 50;
noise = 0.3;
n = 7;
thresh = 0.01;
runs = 10;

% 4)
% Calculate the PM
find_PM_mods_mets(results_location,models_location,mets,metnum,modnum,fixon,fixoff,a,s,samp,noise,n,thresh,runs)

% 5)
% Commented code for calculating PM for 88 essentail biomass components
% across 456 oral microbiome metabolic networks. 
% Running this code will take a long time
% The analysis for Bernstein et al. was conducted on the BU shared
% computing cluster where individual models and metabolites were
% parallelized to improve runtime. To parallelize runs, the metnum and
% modnum variables can be set to variables that change across computers.

% models_location = './metabolic_networks_mat';
% mets = {'cpd00118_c0','cpd00156_c0','cpd00322_c0','cpd00016_c0','cpd00132_c0','cpd00129_c0','cpd00051_c0','cpd00119_c0','cpd02229_c0','cpd00056_c0','cpd00039_c0','cpd00066_c0','cpd00069_c0','cpd00015_c0','cpd00220_c0','cpd00054_c0','cpd00357_c0','cpd00038_c0','cpd00241_c0','cpd00161_c0','cpd00107_c0','cpd00084_c0','cpd00060_c0','cpd00017_c0','cpd11493_c0','cpd00356_c0','cpd00115_c0','cpd00035_c0','cpd00010_c0','cpd00041_c0','cpd00001_c0','cpd00033_c0','cpd00052_c0','cpd00002_c0','cpd00062_c0','cpd00023_c0','cpd00053_c0','cpd00003_c0','cpd00006_c0','cpd00201_c0','cpd00087_c0','cpd00149_c0','cpd00166_c0','cpd15775_c0','cpd15776_c0','cpd15777_c0','cpd15668_c0','cpd15667_c0','cpd15669_c0','cpd15757_c0','cpd15758_c0','cpd15759_c0','cpd00557_c0','cpd11459_c0','cpd15432_c0','cpd00254_c0','cpd00034_c0','cpd00030_c0','cpd00058_c0','cpd00205_c0','cpd00063_c0','cpd15533_c0','cpd15695_c0','cpd15696_c0','cpd00065_c0','cpd00345_c0','cpd00042_c0','cpd15793_c0','cpd15794_c0','cpd15795_c0','cpd15540_c0','cpd15722_c0','cpd15723_c0','cpd00264_c0','cpd00099_c0','cpd00028_c0','cpd15767_c0','cpd15766_c0','cpd15768_c0','cpd15749_c0','cpd15748_c0','cpd15750_c0','cpd10516_c0','cpd10515_c0','cpd15665_c0','cpd15500_c0','cpd15560_c0','cpd15352_c0'};
% metnum = 1:88;
% modnum = 1:456;
% find_PM_mods_mets(results_location,models_location,mets,metnum,modnum,fixon,fixoff,a,s,samp,noise,n,thresh,runs)

**Comment by loretopaulinojr** [comment]
#KBASE HELP, See not working code below. 
# Trying to use the ms2 version of build a model

from biokbase.narrative.jobs.appmanager import AppManager
x = len(names)
for i in range(x):
    genome = names[i]
    model = names[i]+"_modelMS2"
    AppManager().run_app(
        "ModelSEEDReconstruction/build_metabolic_models",
        {
            "genome_refs": genome,
            "suffix": model,
            "run_gapfilling": false,
            "change_to_complete": false,
            "atp_safe": true,
            "gapfilling_media_list": [],
            "forced_atp_list": "",
            "gs_template": "auto",
            "gs_template_ref": [],
            "core_template_ref": [],
        },
        tag="release",
        version="1.0.0"
    )
    
    
 



gapfilling_media_list - text - is a data object where the type is one of: ["KBaseBiochem.Media"]
forced_atp_list - dropdown - must be one of ["Glc.O2", "Ac.O2", "Etho.O2", "Pyr.O2", "Glyc.O2", "Fum.O2", "Succ.O2", "Akg.O2", "LLac.O2", "Dlac.O2", "For.O2", "Glc", "Ac", "Etho", "Pyr", "Glyc", "Fum", "Succ", "Akg", "Llac", "Dlac", "For", "mal-L", "For.NO3", "For.NO2", "For.NO", "Pyr.NO3", "Pyr.NO2", "Pyr.NO", "Ac.NO3", "Ac.NO2", "Ac.NO", "Glc.DMSO", "Pyr.DMSO", "Glc.TMAO", "Pyr.TMAO", "Pyr.SO4", "For.SO4", "LLac.SO4", "Pyr.SO3", "H2.CO2", "H2.Ac", "For.SO4.H2", "LLac.SO4.H2"]
gs_template - dropdown - must be one of ["auto", "old_gramneg", "old_grampos", "gramneg", "grampos", "archaea", "core", "custom"]
gs_template_ref - text - is a data object where the type is one of: ["KBaseFBA.NewModelTemplate"]
core_template_ref - text - is a data object where the type is one of: ["KBaseFBA.NewModelTemplate"]
    
    
    # For Kbase Help, No output came out of this code, this may be becuase I do not have the cell_id and run id
